### ISSUE TITLE
Request: more precise license classifiers

### ISSUE THREAD

**Issue by Steap** [pypa/trove-classifiers#17]
Hello,

Going throught the list of available classifiers at https://pypi.python.org/pypi?%3Aaction=list_classifiers , I feel like some of the license classifiers are not precise enough. For instance, there is a "License :: OSI Approved :: BSD License" that could refer to multiple licenses: BSD-2-Clause, BSD-2-Clause-Patent, BSD-3-Clause. In order to determine the actual license used by a project that only specifies  "License :: OSI Approved :: BSD License", one has to look at the LICENSE file distributed with the source code.

This is an issue for downstream package maintainers for two reasons:
- automated tools (such as pypi2deb, guix import, upt) meant to help them by parsing PyPI and generating a package may have trouble finding the exact license used by a package;
- some versions of a license may be GPL/FSF/DFSG compatible while other versions may not: therefore it makes it harder than necessary to know whether a given package may be included in a given distribution.

I think the following licenses should be added (if possible to both pypi-legacy and warehouse):

- License :: OSI Approved :: Academic Free License 1.1 (AFL-1.1)
- License :: OSI Approved :: Academic Free License 1.2 (AFL-1.2)
- License :: OSI Approved :: Academic Free License 2.0 (AFL-2.0)
- License :: OSI Approved :: Academic Free License 2.1 (AFL-2.1)
- License :: OSI Approved :: Academic Free License 3.0 (AFL-3.0)
- License :: Apache Software License 1.0 (Apache-1.0)
- License :: OSI Approved :: Apache Software License 1.1 (Apache-1.1)
- License :: OSI Approved :: Apache Software License 2.0 (Apache-2.0)
- License :: OSI Approved :: Apple Public Source License 1.0 (APSL-1.0)
- License :: OSI Approved :: Apple Public Source License 1.1 (APSL-1.1)
- License :: OSI Approved :: Apple Public Source License 1.2 (APSL-1.2)
- License :: OSI Approved :: Apple Public Source License 2.0 (APSL-2.0)
- License :: OSI Approved :: Artistic License 1.0 (Artistic-1.0)
- License :: OSI Approved :: Artistic License 2.0 (Artistic-2.0)
- License :: OSI Approved :: BSD 2-Clause "Simplified License" (BSD-2-Clause)
- License :: OSI Approved :: BSD 2-Clause Plus Patent License (BSD-2-Clause-Patent)
- License :: OSI Approved :: BSD 3-Clause "New" or "Revised" License (BSD-3-Clause)
- License :: OSI Approved :: GNU Lesser General Public License v2.0 (LGPLv2.0)
- License :: OSI Approved :: GNU Lesser General Public License v2.0 or later (LGPLv2.0+)
- License :: OSI Approved :: GNU Lesser General Public License v2.1 (LGPLv2.1)
- License :: OSI Approved :: GNU Lesser General Public License v2.1 or later (LGPLv2.1+)
- License :: OSI Approved :: GNU Lesser General Public License v3.0 (LGPLv3.0)
- License :: OSI Approved :: GNU Lesser General Public License v3.0 or later (LGPLv3.0+)

In parentheses are the spdx identifiers (see https://spdx.org/licenses/) except for LGPL* where I used identifiers similar to those currently used for the various versions of the GPL.

Regarding the LGPL classifiers, we may also state that v2 and v2+ (currently in the list of valid classifiers) refer to v2.0 and v2.0+ and **not** to v2.1 and v2.1+, which would remove the need for the LGPLv2 and LGPLv2.0+ classifiers.

I decided not to include less used variants of the BSD licences - they may be added in the future if need be.

What do you think about this?

**Comment by mschwager** [comment]
Has there been any progress here? I'm still waiting on `LGPLv2.1`.

**Comment by PuneetGopinath** [comment]
> Has there been any progress here? I'm still waiting on `LGPLv2.1`.

I'm also still waiting on `LGPLv2.1`!!

**Comment by Alireza1369newi** [comment]
> This issue is now unblocked. After compiling all the differences between the following sources:
> * https://spdx.org/licenses/
> * https://opensource.org/licenses
> * https://pypi.org/classifiers/
> 
> I think this is what needs done (red will be deprecated, green will be added):
> 
> ### Academic
> ```diff
> - License :: OSI Approved :: Academic Free License (AFL)
> + License :: OSI Approved :: Academic Free License 1.1 (AFL-1.1)
> + License :: OSI Approved :: Academic Free License 1.2 (AFL-1.2)
> + License :: OSI Approved :: Academic Free License 2.0 (AFL-2.0)
> + License :: OSI Approved :: Academic Free License 2.1 (AFL-2.1)
> + License :: OSI Approved :: Academic Free License 3.0 (AFL-3.0)
> ```
> 
> ### Apache
> ```diff
> - License :: OSI Approved :: Apache Software License
> + License :: Apache License, Version 1.1 (Apache-1.1)
> + License :: Apache License, Version 1.0 (Apache-1.0)
> + License :: OSI Approved :: Apache Software License 2.0 (Apache-2.0)
> ```
> 
> ### Apple
> ```diff
> - License :: OSI Approved :: Apple Public Source License
> + License :: OSI Approved :: Apple Public Source License 1.0 (APSL-1.0)
> + License :: OSI Approved :: Apple Public Source License 1.1 (APSL-1.1)
> + License :: OSI Approved :: Apple Public Source License 1.2 (APSL-1.2)
> + License :: OSI Approved :: Apple Public Source License 2.0 (APSL-2.0)
> ```
> 
> ### Artistic
> ```diff
> - License :: OSI Approved :: Artistic License
> + License :: OSI Approved :: Artistic License 1.0 (Artistic-1.0)
> + License :: OSI Approved :: Artistic License 2.0 (Artistic-2.0)
> ```
> 
> ### BSD
> ```diff
> - License :: OSI Approved :: BSD License
> + License :: OSI Approved :: BSD 2-Clause Plus Patent License (BSD-2-Clause-Patent)
> + License :: OSI Approved :: BSD 2-Clause "Simplified" License (BSD-2-Clause)
> + License :: OSI Approved :: BSD 3-Clause "New" or "Revised" License (BSD-3-Clause)
> ```
> 
> ### GNU
> ```diff
> - License :: OSI Approved :: GNU Affero General Public License v3
> + License :: OSI Approved :: GNU Affero General Public License v3.0 only (AGPL-3.0-only)
> - License :: OSI Approved :: GNU Affero General Public License v3 or later (AGPLv3+)  
> + License :: OSI Approved :: GNU Affero General Public License v3.0 or later (AGPL-3.0-or-later)
> - License :: OSI Approved :: GNU Free Documentation License (FDL)
> + License :: OSI Approved :: GNU Free Documentation License v1.1 only (GFDL-1.1-only)
> + License :: OSI Approved :: GNU Free Documentation License v1.1 or later (GFDL-1.1-or-later)
> + License :: OSI Approved :: GNU Free Documentation License v1.2 only (GFDL-1.2-only)
> + License :: OSI Approved :: GNU Free Documentation License v1.2 or later (GFDL-1.2-or-later)
> + License :: OSI Approved :: GNU Free Documentation License v1.3 only (GFDL-1.3-only)
> + License :: OSI Approved :: GNU Free Documentation License v1.3 or later (GFDL-1.3-or-later)
> - License :: OSI Approved :: GNU General Public License (GPL)
> + License :: GNU General Public License v1.0 only (GPL-1.0-only)
> + License :: GNU General Public License v1.0 or later (GPL-1.0-or-later)
> - License :: OSI Approved :: GNU General Public License v2 (GPLv2)
> + License :: OSI Approved :: GNU General Public License v2.0 only (GPL-2.0-only)
> - License :: OSI Approved :: GNU General Public License v2 or later (GPLv2+)
> + License :: OSI Approved :: GNU General Public License v2.0 or later (GPL-2.0-or-later)
> - License :: OSI Approved :: GNU General Public License v3 (GPLv3)
> + License :: OSI Approved :: GNU General Public License v3.0 only (GPL-3.0-only)
> - License :: OSI Approved :: GNU General Public License v3 or later (GPLv3+)  
> + License :: OSI Approved :: GNU General Public License v3.0 or later (GPL-3.0-or-later)
> - License :: OSI Approved :: GNU Lesser General Public License v2 (LGPLv2)
> + License :: OSI Approved :: GNU Library General Public License v2 only (LGPL-2.0-only)
> - License :: OSI Approved :: GNU Lesser General Public License v2 or later (LGPLv2+)
> + License :: OSI Approved :: GNU Library General Public License v2 or later (LGPL-2.0-or-later)
> + License :: OSI Approved :: GNU Lesser General Public License v2.1 only (LGPL-2.1-only)
> + License :: OSI Approved :: GNU Lesser General Public License v2.1 or later (LGPL-2.1-or-later)
> - License :: OSI Approved :: GNU Lesser General Public License v3 (LGPLv3)
> + License :: OSI Approved :: GNU Lesser General Public License v3.0 only (LGPL-3.0-only)
> - License :: OSI Approved :: GNU Lesser General Public License v3 or later (LGPLv3+)
> + License :: OSI Approved :: GNU Lesser General Public License v3.0 or later (LGPL-3.0-or-later)
> - License :: OSI Approved :: GNU Library or Lesser General Public License (LGPL)
> ```
> 
> The deprecated classifiers will affect a lot of projects in some cases:
> 
> | Classifier | # of projects |
> |--|--:|
> | `Academic Free License (AFL)` | 40 |
> | `Apache Software License` | 8176 |
> | `Apple Public Source License` | 4 |
> | `Artistic License` | 46 |
> | `BSD License` | >10000 |
> | `GNU Affero General Public License v3` | 3859 |
> | `GNU Affero General Public License v3 or later (AGPLv3+) ` | 524 |
> | `GNU Free Documentation License (FDL)` | 3 |
> | `GNU General Public License (GPL)` | 3794 |
> | `GNU General Public License v2 (GPLv2)` | 1152 |
> | `GNU General Public License v2 or later (GPLv2+)` | 408 |
> | `GNU General Public License v3 (GPLv3)` | 3124 |
> | `GNU General Public License v3 or later (GPLv3+) ` | 2081 |
> | `GNU Lesser General Public License v2 (LGPLv2)` | 128 |
> | `GNU Lesser General Public License v2 or later (LGPLv2+)` | 175 |
> | `GNU Lesser General Public License v3 (LGPLv3)` | 831 |
> | `GNU Lesser General Public License v3 or later (LGPLv3+)` | 502 |
> | `GNU Library or Lesser General Public License (LGPL)` | 1194 |
> 
> @Steap, can you review?
> 
> @dstufft @ewdurbin Can you sanity-check? Currently a user attempting to publish a release with a deprecated classifier will get an error like:
> 
> ```
> HTTPError: 400 Client Error: Invalid value for classifiers. Error: Classifier
> 'Topic :: Communications :: Chat :: AOL Instant Messenger' has been deprecated,
> see https://pypi.org/classifiers/ for a list of valid classifiers. for url:
> http://upload.pypi.org/legacy/
> ```

**Comment by edgarrmondragon** [comment]
Does [PEP 639](https://peps.python.org/pep-0639/) make this unnecessary?

**Comment by di** [comment]
Yes, as a result of PEP 639, we won't add any new license classifiers here, but I think we should probably explore deprecating all license classifiers instead, eventually.

This largely depends on the state of various build backends and what they do when they encounter license classifiers (e.g. setuptools emits a deprecation warning https://github.com/pypa/setuptools/pull/4706/ and avoids writing them to the metadata file https://github.com/pypa/setuptools/pull/4833).

**Comment by di** [comment]
Closing in favor of https://github.com/pypa/trove-classifiers/issues/236.
